### PR TITLE
New Vehicle Heron USV for Webots 

### PR DESCRIPTION
**Pull request by harunkurtdev**
heron_ocean

By connecting heron usv to ardupilot on webots.
A new vehicle type has been added. This vehicle type is a marine vehicle and is well known in the market.

### CODE REVIEW

**Comment by rmackay9** [comment]
Hi @harunkurtdev,

Thanks for this.  I actually didn't realise that Webots was still usable.  Good news that it is.

One thing is it would be best if the commits were squashed together somewhat and also if you peek at our commit history you'll see we always prefix commits with the subsystem (or folder) affected.  This makes it easier for us to backport features.

**Comment by harunkurtdev** [comment]
> Hi @harunkurtdev,
> 
> Thanks for this. I actually didn't realise that Webots was still usable. Good news that it is.
> 
> One thing is it would be best if the commits were squashed together somewhat and also if you peek at our commit history you'll see we always prefix commits with the subsystem (or folder) affected. This makes it easier for us to backport features.

So sir, what should I do? Can you recommend me a guideline?

I look this [contributing](https://github.com/ArduPilot/ardupilot/blob/637aec00853015eaaccab40217a6338164cffce1/.github/CONTRIBUTING.md)

**Comment by harunkurtdev** [comment]
Mr. @rmackay9, it is working,  what should I do now? I did not add any file confusion that would prevent Ardupilot from working.

**Comment by peterbarker** [comment]
> > Hi @harunkurtdev,
> > Thanks for this. I actually didn't realise that Webots was still usable. Good news that it is.
> > One thing is it would be best if the commits were squashed together somewhat and also if you peek at our commit history you'll see we always prefix commits with the subsystem (or folder) affected. This makes it easier for us to backport features.
> 
> So sir, what should I do? Can you recommend me a guideline?
> 
> I look this [contributing](https://github.com/ArduPilot/ardupilot/blob/637aec00853015eaaccab40217a6338164cffce1/.github/CONTRIBUTING.md)

The information you're after is here: https://ardupilot.org/dev/docs/git-interactive-rebase.html

File: libraries/SITL/SIM_Webots_Python.cpp [hunk 1]
```diff
@@ -45,6 +48,7 @@ void WebotsPython::set_interface_ports(const char* address, const int port_in, c
     // try to bind to a specific port so that if we restart ArduPilot
     // Webots keeps sending us packets. Not strictly necessary but
     // useful for debugging
+
```
Comment (peterbarker) [inline libraries/SITL/SIM_Webots_Python.cpp:51]: Bad whitespace change

File: libraries/SITL/SIM_Webots_Python.cpp [hunk 2]
```diff
@@ -53,8 +57,17 @@ void WebotsPython::set_interface_ports(const char* address, const int port_in, c
...
+    
+    /*
+    
+    if (is_wsl()) {
+        _webots_address = "0.0.0.0"; // Use 0.0.0.0 in WSL
+    } else {
+        _webots_address = address; // Use localhost otherwise
+    }
+    */
+
```
Comment (peterbarker) [inline libraries/SITL/SIM_Webots_Python.cpp:69]: Removed change?

File: libraries/SITL/SIM_Webots_Python.cpp [hunk 3]
```diff
@@ -128,6 +141,21 @@ void WebotsPython::recv_fdm(const struct sitl_input &input)
 
 }
 
+bool WebotsPython::is_wsl(){
```
Comment (peterbarker) [inline libraries/SITL/SIM_Webots_Python.cpp:144]: Unused?

File: libraries/SITL/SIM_Webots_Python.h [hunk 4]
```diff
@@ -29,6 +29,9 @@
 #include "SIM_Aircraft.h"
 #include <AP_HAL/utility/Socket_native.h>
 
+
+
+
```
Comment (peterbarker) [inline libraries/SITL/SIM_Webots_Python.h:34]: Bad whitespace change

File: libraries/SITL/SIM_Webots_Python.h [hunk 5]
```diff
@@ -67,6 +70,10 @@ class WebotsPython : public Aircraft {
       double position_xyz[3];
     };
 
+
+    /// @brief check for wsl env
+    bool is_wsl();
```
Comment (peterbarker) [inline libraries/SITL/SIM_Webots_Python.h:75]: Unused?

File: libraries/SITL/examples/Webots_Python/controllers/ardupilot_vehicle_controller/webots_vehicle.py [hunk 6]
```diff
@@ -155,14 +155,13 @@ def _handle_sitl(self, sitl_address: str = "127.0.0.1", port: int = 9002):
         Args:
             port (int, optional): Port to listen for SITL on. Defaults to 9002.
         """
-
```
Comment (peterbarker) [inline libraries/SITL/examples/Webots_Python/controllers/ardupilot_vehicle_controller/webots_vehicle.py:158]: bad whitespace change?

File: libraries/SITL/examples/Webots_Python/controllers/ardupilot_vehicle_controller/webots_vehicle.py [hunk 7]
```diff
@@ -155,14 +155,13 @@ def _handle_sitl(self, sitl_address: str = "127.0.0.1", port: int = 9002):
...
         # create a local UDP socket server to listen for SITL
         s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM) # SOCK_STREAM
         s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
         s.bind(('0.0.0.0', port))
 
         # wait for SITL to connect
-        print(f"Listening for ardupilot SITL (I{self._instance}) at 127.0.0.1:{port}")
+        print(f"Listening for ardupilot SITL (I{self._instance}) at {sitl_address}:{port}")
```
Comment (peterbarker) [inline libraries/SITL/examples/Webots_Python/controllers/ardupilot_vehicle_controller/webots_vehicle.py:164]: Didn't we just bind 0.0.0.0?

I mean, the existing comment is wrong, but this one is also wrong.

I mean, *less* wrong :-)

File: libraries/SITL/examples/Webots_Python/controllers/ardupilot_vehicle_controller/webots_vehicle.py [hunk 8]
```diff
@@ -241,12 +240,13 @@ def _handle_controls(self, command: tuple):
         Args:
             command (tuple): tuple of motor speeds 0.0-1.0 where -1.0 is unused
         """
-
+
```
Comment (peterbarker) [inline libraries/SITL/examples/Webots_Python/controllers/ardupilot_vehicle_controller/webots_vehicle.py:243]: bad whitespace change

File: libraries/SITL/examples/Webots_Python/controllers/ardupilot_vehicle_controller/webots_vehicle.py [hunk 9]
```diff
@@ -241,12 +240,13 @@ def _handle_controls(self, command: tuple):
...
             command (tuple): tuple of motor speeds 0.0-1.0 where -1.0 is unused
         """
-
+        
         # get only the number of motors we have
         command_motors = command[:len(self._motors)]
         if -1 in command_motors:
             print(f"Warning: SITL provided {command.index(-1)} motors "
                   f"but model specifies {len(self._motors)} (I{self._instance})")
+            command_motors = [v if v != -1 else 0.0 for v in command_motors]
```
Comment (peterbarker) [inline libraries/SITL/examples/Webots_Python/controllers/ardupilot_vehicle_controller/webots_vehicle.py:249]: Add a comment as to why this is needed, please.

File: libraries/SITL/examples/Webots_Python/params/heron.parm [hunk 10]
```diff
@@ -0,0 +1,13 @@
+FRAME_CLASS      2          # Boat
+
+SERVO1_FUNCTION  73         # Throttle Left (Boat throttle channel)
+SERVO2_FUNCTION  74         # Throttle Right (Boat throttle channel)
+SERVO3_FUNCTION  0          # Disabled (no steering required)
+SERVO4_FUNCTION  0          # Disabled (no steering required)
+
+AHRS_EKF_TYPE    10         # Use sim AHRS
+#ARMING_CHECK	 1045534    # No RC requirements
```
Comment (peterbarker) [inline libraries/SITL/examples/Webots_Python/params/heron.parm:9]: Why remove RC here?  Simulation has it by default?

If this was copied from elsewhere that's fine.

**Review by peterbarker — CHANGES_REQUESTED** [review]

**Comment by peterbarker** [comment]
@harunkurtdev still waiting for some of my comments to be addressed

**Comment by harunkurtdev** [comment]
> @harunkurtdev still waiting for some of my comments to be addressed

Mr. @peterbarker,

 I removed the parts you took into consideration and the unused parts.

Which part do I need to edit? Could you please state it again?

**Review by peterbarker — CHANGES_REQUESTED** [review]
Ping @harunkurtdev - there are many un-addressed review comments in here.

**Pull request closed without merging**